keps/sig-network/3458-remove-transient-node-predicates-from-service-controller/README.md

```diff
@@ -0,0 +1,376 @@
+
+## Motivation
+
+Abruptly terminating connections in the cases defined by a) and b) above can be
+seen as buggy behavior and should be improved. By enabling connection draining,
+applications are allowed profit from graceful shutdown / termination, for what
+concerns cluster ingress connectivity. Users of Kubernetes will also see a
+reduction in the amount of cloud API calls, for what concerns calls stemming
+from syncing load balancers with the Kubernetes cluster state.
+
+### Goals
+
+- Stop re-configuring the load balancers' node set for cases a) and b) above
+
+### Non-Goals
+
+- Stop re-configuring the load balancers' node set for fully deleted /
+  newly added cluster nodes, or for nodes which get annotated with
+  `node.kubernetes.io/exclude-from-external-load-balancers`.
+- Enable load balancer connection draining while Node is draining. This requires
+  health check changes.
+
+## Proposal
+
+### Risks and Mitigations
+
+#### Risk
+
+1. Cloud providers which do not allow VM deletion when the VM is referenced by
+  other constructs, will block the cluster auto-scaler (CA) from deleting the VM
+  upon downscale. This will result in reduced downscale performance by the CA,
+  or completely block VM deletion from happening - this is because the service
+  controller will never proceed to de-reference the VM from the load balancer
+  node set until the Node is fully deleted in the API server, which will never
+  occur until the VM is deleted. The three major cloud providers (GCP/AWS/Azure)
+  do however support this, and it is not expected that other providers don't.
+2. Cloud providers which do not configure their load balancer health checks to
+  target the service proxy's healthz, alternatively: constructs which validate
+  the endpoint's reachability across the data plane; risk experiencing
+  regressions as a consequence of the removal of b). This would happen if a node
+  is faced with a terminal error which does impact the Node's network
+  connectivity. Doing this is considered incorrect, and therefor not expected to
+  be the case.
+3. By removing b) above we are delaying the removal of the Node from the load
+  balancers' node set until the Node is completely deleted in the API server.
+  This might have an impact on CA downscaling. The reason for this is: the CA
+  deletes the VM and expects the node controller in the KCCM to notice this and
+  delete the Node in Kubernetes, as a consequence. If the node controller takes
+  a while to sync that and other Node related events trigger load balancer
+  reconciliation while this is happening, then the service controller will error
+  until the cluster reaches steady-state (because it's trying to sync Nodes for
+  which the VM is non-existent). It is therefor encouraged that cloud providers
+  check the existence of the underlying VM corresponding to each Node, when
+  syncing load balancers. Due to the removal of b): syncing load balancers as a
```

Comment:
Relocated discussion from elsewhere.

If CA taints a Node, then deletes the VM, and NodeController is slow (say it takes 60 seconds to wake up), any subsequent LB set changes will include this Node, but the VM for it is already gone, so will fail. 

You responded that this could get expensive, checking `InstanceExists()` for every VM.

Does this argue for https://github.com/kubernetes/enhancements/pull/3460/files#r1093923670 - not adding a node which is unready or not adding if the ToBeDeleted taint exists?  If the NodeController is MIA, then we can't really count on Node ready->unready happening, but we know that CA adds that taint BEFORE deleting the VM, so we can rely on it.

It adds logic complexity, but avoids this problem (at least in the case of CA, which is signficant).

Reply:
> not adding a node which is unready or not adding if the ToBeDeleted taint exists? 

OK, so we are essentially saying that we should not enqueue node updates on any change except for "exclude". When we sync an enqueued event however: we only list all Nodes which are ready and not tainted. Shouldn't be that difficult in terms of complexity....but it goes against the entire goal of the KEP. If we end up implementing this we only change one single line of code in the service controller, [removing this one](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/service/controller.go#L674) :)

I kept this part in the `Risk`, but added this solution to `Mitigation`. Hope that's fine?

Reply:
I don't think that is QUITE what I am saying.

![Node LB states, v4 (KEP 1)](https://user-images.githubusercontent.com/5595220/216483238-a93fd90b-ce80-46ea-b1da-ebc7d7167c9e.png)

Reply:
Then KEP 2 is something like:
![Node LB states, v4 (KEPs 1 and 2)](https://user-images.githubusercontent.com/5595220/216483320-43a6bf7c-0f01-4bb2-9e14-c88082b9c978.png)

Reply:
> but worth evaluating? 

Of course. 

I need to validate the cloud-provider code though...because my statement:

> checking InstanceExists() for every VM.

shouldn't really need to be true. The service controller passes the `Node` object to the cloud provider code. The `Node` object encodes `providerID`. The load balancer references an instance ID which is that `providerID`. If the cloud provider code just makes sure to implement "normal" reconciliation logic, i.e: for every load balancer update: list all referenced instance IDs, compare to the `providerID`, if `providerID` not in referenced IDs then dequeue (`return nil` - no work to do).

That would save the service controller from complex logic and solve the problem without any cloud API calls.

Reply:
And if that isn't the case, then I could update all in-tree cloud providers to do that....out-of-tree ones though 🤷

Reply:
Actually what I said doesn't work: the cloud provider code has no clue what type of an operation we're dealing with, the service controller just calls: `EnsureLoadBalancer(service, nodes...)` so if an instance ID is not in the set, then it adds it.

Reply:
Revised drawing.

![Node LB states, v5 (KEP 1)](https://user-images.githubusercontent.com/5595220/216734146-d3e989f5-af62-4948-90c9-eb72160aa2fc.png)